Modify RealtimeMarkingScheme for moving to OMR

jason-hall opened the pull request:
Signed-off-by: Jason Hall <jasonhal@ca.ibm.com>

jason-hall commented:
I moved the classes inheriting from RealtimeRootScanner from RealtimeMarkingScheme into RealtimeRootScanner. The intention here is to put RealtimeRootScanner into the glue folder where the other such Scanners live. I also added delegate functions for the java-specific marking code in RealtimeMarkingScheme; OMR needs much simpler implementations

dmitripivkine commented:
```
../gc_realtime/libj9realtime.a(RealtimeMarkingScheme.cpp.o): In function `MM_RealtimeMarkingScheme::incrementalConsumeQueue(MM_EnvironmentRealtime*, unsigned long)':
/home/travis/build/eclipse/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/vm/runtime/gc_realtime/../../../../../openj9/runtime/gc_realtime/RealtimeMarkingScheme.cpp:115: undefined reference to `MM_MetronomeDelegate::scanPointerArraylet(MM_EnvironmentRealtime*, unsigned int*)'
/home/travis/build/eclipse/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/vm/runtime/gc_realtime/../../../../../openj9/runtime/gc_realtime/RealtimeMarkingScheme.cpp:118: undefined reference to `MM_MetronomeDelegate::scanObject(MM_EnvironmentRealtime*, J9Object*)'
collect2: error: ld returned 1 exit status
make[6]: *** [runtime/libj9gc29.so] Error 1
make[5]: *** [runtime/gc/CMakeFiles/j9gc.dir/all] Error 2
```

jason-hall commented:
@dmitripivkine I don't understand it. It compiles fine for me and that error doesn't make any sense. The prototype for the function it says is undefined isn't even in the code anywhere. Its a different function prototype. I can't see how that error is possible. Could it be trying to compile an older version of one of the files somehow?

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented:
Failed to compile on Windows as well:
```
21:22:09     Creating library ../lib/j9gc.lib and object ../lib/j9gc.exp
21:22:09  ../makelib/targets.mk:459: UMA_SINGLE_REBASE specified, suppressing per-directory rebase.
21:22:09  j9realtime.lib(RealtimeMarkingScheme.obj) : error LNK2019: unresolved external symbol "public: unsigned __int64 __cdecl MM_MetronomeDelegate::scanObject(class MM_EnvironmentRealtime *,struct J9Object *)" (?scanObject@MM_MetronomeDelegate@@QEAA_KPEAVMM_EnvironmentRealtime@@PEAUJ9Object@@@Z) referenced in function "public: bool __cdecl MM_RealtimeMarkingScheme::incrementalConsumeQueue(class MM_EnvironmentRealtime *,unsigned __int64)" (?incrementalConsumeQueue@MM_RealtimeMarkingScheme@@QEAA_NPEAVMM_EnvironmentRealtime@@_K@Z)
21:22:09  j9realtime.lib(RealtimeMarkingScheme.obj) : error LNK2019: unresolved external symbol "public: unsigned __int64 __cdecl MM_MetronomeDelegate::scanPointerArraylet(class MM_EnvironmentRealtime *,unsigned int *)" (?scanPointerArraylet@MM_MetronomeDelegate@@QEAA_KPEAVMM_EnvironmentRealtime@@PEAI@Z) referenced in function "public: bool __cdecl MM_RealtimeMarkingScheme::incrementalConsumeQueue(class MM_EnvironmentRealtime *,unsigned __int64)" (?incrementalConsumeQueue@MM_RealtimeMarkingScheme@@QEAA_NPEAVMM_EnvironmentRealtime@@_K@Z)
21:22:09  ../j9gc29.dll : fatal error LNK1120: 2 unresolved externals
21:22:09  make[3]: *** [../makelib/targets.mk:458: ../j9gc29.dll] Error 96
21:22:09  make[2]: *** [makefile:1028: j9gc] Error 2
```

dmitripivkine commented on runtime/gc_realtime/RealtimeMarkingScheme.cpp:
```diff
@@ -20,405 +20,12 @@
-#include "StackSlotValidator.hpp"
-#include "UnfinalizedObjectBufferRealtime.hpp"
-#include "UnfinalizedObjectList.hpp"
-#include "WorkPacketsRealtime.hpp"
```
this include still required

dmitripivkine replied:
...or not. Probably this comment is not correct. Let me double check

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented on runtime/gc_realtime/RealtimeMarkingScheme.hpp:
```diff
@@ -51,8 +50,7 @@ class MM_RealtimeMarkingScheme : public MM_SegregatedMarkingScheme
 	MM_Scheduler *_scheduler;
-	J9JavaVM *_javaVM;  /**< The current JVM*/
-	MM_GCExtensions *_gcExtensions;
+	MM_GCExtensionsBase *_gcExtensions;
```
looks like this variable is not necessary. The hierarchy is:
`MM_RealtimeMarkingScheme` <- `MM_SegregatedMarkingScheme` <- `MM_MarkingScheme`
and `MM_MarkingScheme` has `MM_GCExtensionsBase *_extensions;` already. It can be used instead

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1130,712 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
 
+UDATA
+MM_MetronomeDelegate::scanPointerRange(MM_EnvironmentRealtime *env, fj9object_t *startScanPtr, fj9object_t *endScanPtr)
+{
```
This method use to be inlined too. Can it be moved to .hpp as well?

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1130,712 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+}
+
+UDATA
+MM_MetronomeDelegate::scanMixedObject(MM_EnvironmentRealtime *env, J9Object *objectPtr)
```
This method use to be inlined too. Can it be moved to .hpp as well?

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1130,712 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+}
+
+UDATA
+MM_MetronomeDelegate::scanReferenceMixedObject(MM_EnvironmentRealtime *env, J9Object *objectPtr)
```
This method use to be inlined too. Can it be moved to .hpp as well?

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1130,712 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+}
+
+UDATA
+MM_MetronomeDelegate::scanPointerArrayObject(MM_EnvironmentRealtime *env, J9IndexableObject *objectPtr)
```
This method use to be inlined too. Can it be moved to .hpp as well?

jason-hall commented:
I was able to move those functions into the MetronomeDelegate header by making the following changes.
- The functions all use functions from MM_EnvironmentRealtime. EnvironmentRealtime.hpp includes Scheduler.hpp which includes MetronomeDelegate.hpp. I removed the Scheduler include from EnvironmentRealtime.hpp
- The functions use MM_RealtimeGC->MM_RealtimeMarkingScheme. I added a MM_RealtimeMarkingScheme* field to MetronomeDelegate to fix this
- MM_RealtimeMarkingScheme header used a macro defined in RealtimeGC.hpp, which I moved to RealtimeMarkingScheme.hpp
- Moved MetronomeDelegate include from Scheduler.hpp to Scheduler.cpp. It was unnecessary and now MetronomeDelegate header can include Scheduler
- Moved the one line of code from RealtimeGC::enqueuePointerArraylet (used nowhere else) to where it is used in the delegate

jason-hall commented:
I rushed these in so you can look at them if you want. I'll cone back to this after my appointment in the afternoon

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.hpp:
```diff
@@ -157,10 +166,374 @@ class MM_MetronomeDelegate : public MM_BaseNonVirtual
+							_markingScheme->setScanAtomic((J9Object *)objectPtr);
+						}
+					} else {
+						env->getWorkStack()->push(env, startScanPtr);
```
There was call to `MM_RealtimeGC::enqueuePointerArraylet()` here. This function not only push object pointer to Work Stack but use `ARRAYLET_TO_ITEM()` macro to add `ITEM_IS_ARRAYLET` flag to it. Am I correct that this logic is missed here? Looks like it should be
```
env->getWorkStack()->push(env, ARRAYLET_TO_ITEM(startScanPtr));
```

dmitripivkine commented on runtime/gc_realtime/RealtimeRootScanner.hpp:
```diff
@@ -117,5 +118,377 @@ class MM_RealtimeRootScanner : public MM_RootScanner
+	scanOneThreadImpl(MM_EnvironmentRealtime *env, J9VMThread* walkThread, void* localData)
+	{
+		MM_EnvironmentRealtime* walkThreadEnv = MM_EnvironmentRealtime::getEnvironment(walkThread->omrVMThread);
+		MM_RealtimeGC* realtimeGC = (MM_RealtimeGC*)_realtimeGC;
```
The cast is not necessary because `_realtimeGC` has type `(MM_RealtimeGC*)` already. I would not declare `realtimeGC` at all but use `_realtimeGC` instead.

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1131,430 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+		 */
+		if(isDynamicClassUnloadingEnabled()) {
+			((J9ClassLoader *)_javaVM->systemClassLoader)->gcFlags |= J9_GC_CLASS_LOADER_SCANNED;
+			_realtimeGC->getMarkingScheme()->markObject(env, (J9Object *)((J9ClassLoader *)_javaVM->systemClassLoader)->classLoaderObject);
```
`_realtimeGC->getMarkingScheme()` can be replaced to `_markingScheme` everywhere in `MM_MetronomeDelegate`

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.hpp:
```diff
@@ -157,10 +166,374 @@ class MM_MetronomeDelegate : public MM_BaseNonVirtual
+	{
+		fj9object_t *startScanPtr = arraylet;
+		fj9object_t *endScanPtr = startScanPtr + (_javaVM->arrayletLeafSize / sizeof(fj9object_t));
+		return scanPointerRange(env, startScanPtr, endScanPtr);
```
Style: mixing of `fomrobject_t` and `fj9object_t` in the same function looks inconsistent. Considering transition to OMR types I would replace  `fj9object_t` to `fomrobject_t` in this function

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.hpp:
```diff
@@ -157,10 +166,374 @@ class MM_MetronomeDelegate : public MM_BaseNonVirtual
 
+	void markLiveObjects(MM_EnvironmentRealtime *env);
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+	MMINLINE void markClassOfObject(MM_EnvironmentRealtime *env, J9Object *objectPtr)
```
Style - new look: for functions declared in .hpp files we should use the same style as in .cpp - have declarated types in separate line (at least for new code):
```
MMINLINE void 
markClassOfObject(MM_EnvironmentRealtime *env, J9Object *objectPtr)
```
If it is convenient for you, would you please use this format for this file?

dmitripivkine commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1131,430 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+		objectsVisited += 1;
+		referenceStats->_candidates += 1;
+
+		Assert_MM_true(_realtimeGC->_markingScheme->isMarked(referenceObj));
```
also can be replaced to _markingScheme

dmitripivkine commented:
Jenkins compile win

amicic commented:
There is a lot of code that goes into Delegate side that should stay in non-delegate side (and eventually move to OMR). 

The prime example is the high level method markLiveObjects(). The whole method should actually reside in non-delegate side, and only two calls to root scanner (hard roots before and weak roots processing after the main tracing loop) should rely on delegate. There are more spots/methods, where a finer line between delegate and non-delegate code could be drawn. 

MarkingScheme in OMR side (used for Standard group of collectors) is good model to follow.

jason-hall commented:
@amicic I moved all of the markLiveObjects() logic back into RealtimeMarkingScheme as suggested. I agree its a good idea. All other changes here are just the java-specific scanning code (markClass(), scanReferenceMixedObject(), etc).

OMR's MarkingScheme markLiveObjects functions seem to have no logic in them and just directly call the delegate. Do you have any other specific suggestions on other spots/methods?

amicic commented:
I was under impression that some of indexable code (splitting logic, for example) could move too, but I can see now that even for Standard collectors it resides in Delegate side (or Scanner to be more specific). I guess, in general, i'd like to see that some of it move to OMR, but it's out of scope of this PR.

A suggestion for a few methods renaming to match with Standard GCs:
- doTracing in Realtime very resembles to completeMarking in MarkingScheme
- incrementalConsumeQueue resembles to completeScan, but we can call it incrementalCompleteScan
- markLiveObject could be broken down into: markLiveObjectsInit, markLiveObjectsRoots, markLiveObjectsScan, markLiveObjectsComplete

amicic commented:
Another minor suggestion, which does not fit into delegate discussion, but it's complying with my suggestion to align code with OMR MarkingScheme as much as possible.

Move some of code from MM_RealtimeMarkTask::setup/cleanup to markLiveObjectsInit/markLiveObjectsComplete.

Maybe there is a chance that MM_RealtimeMarkTask can completely go away and use MM_ParallelMarkTask instead

jason-hall commented:
I updated the function names to your better suggestions, but I don't think MM_RealtimeMarkTask can be easily replaced with MM_ParallelMarkTask. MM_RealtimeMarkTask is one of two classes inheriting from MM_IncrementalParallelTask which adds a whole bunch of things.

dmitripivkine commented on runtime/gc_realtime/RealtimeRootScanner.hpp:
```diff
@@ -117,5 +118,376 @@ class MM_RealtimeRootScanner : public MM_RootScanner
+/**
+ * This scanner will mark objects that pass through its doSlot.
+ */
+class MM_RealtimeMarkingSchemeRootMarker : public MM_RealtimeRootScanner
```
Would you please create RealtimeMarkingSchemeRootMarker.hpp (and RealtimeMarkingSchemeRootMarker.cpp if necessary) and move this code there?

dmitripivkine commented on runtime/gc_realtime/RealtimeRootScanner.hpp:
```diff
@@ -117,5 +118,376 @@ class MM_RealtimeRootScanner : public MM_RootScanner
+/**
+ * This scanner will mark objects that pass through its doSlot.
+ */
+class MM_RealtimeMarkingSchemeRootClearer : public MM_RealtimeRootScanner
```
Would you please create MM_RealtimeMarkingSchemeRootClearer.hpp (and MM_RealtimeMarkingSchemeRootClearer.cpp if necessary) and move this code there?

jason-hall commented:
@dmitripivkine @amicic Is the latest update an improvement? The diff looks messy for some reason, but I divided the existing markLiveObjects() into those functions without changing the content

amicic commented on runtime/gc_realtime/RealtimeMarkingScheme.cpp:
```diff
@@ -480,502 +83,88 @@ void
+	metronomeDelegate->markLiveObjectsRoots(realtimeEnv);
+	_scheduler->condYieldFromGC(realtimeEnv);
+	/* Heap Marking and barrier processing. Cannot delay barrier processing until the end.*/	
+	_realtimeGC->completeMarking(realtimeEnv);
```
This is where markLiveObjectsScan should start (including this line).

amicic replied:
Again the model is base MarkingScheme (non-realtime). Over there markLiveObjectsScan, does completeMarking, what is the minimum core work. RT does a bit more.

jason-hall replied:
Okay, that's a simple fix. Done

amicic commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1133,396 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+}
+
+void
+MM_MetronomeDelegate::flushRememberedSet(MM_EnvironmentRealtime *env)
```
This is not remember set. 
This should be renamed to flushReferenceObjects (and possibly called from delegate's markLiveObjectsScan - not sure where exactly markLiveObjectsScan is called from)

jason-hall replied:
Moving it to the delegate function would change the order of the code from its original implementation. possibly breaking it. markLiveObjectsScan() is called from the spot you showed me the other day in RealtimeMarkTask.

amicic commented on runtime/gc_glue_java/MetronomeDelegate.cpp:
```diff
@@ -1115,5 +1133,396 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+void
+MM_MetronomeDelegate::markLiveObjectsScan(MM_EnvironmentRealtime *env)
+{
+	MM_RealtimeMarkingSchemeRootClearer rootScanner(env, _realtimeGC);
```
Does not look right. Clearable roots scanning belongs to complete phase. Again, look at the base MarkingScheme.

jason-hall replied:
Again, if you look at the base class, it has no such code in the complete phase. There are no clearable roots in any phase. I'm not sure why you think its an example I can follow. Perhaps you are thinking of another class somewhere else entirely.

void MM_MarkingScheme::markLiveObjectsComplete(MM_EnvironmentBase *env)
{
	_delegate.workerCompleteGC(env);
}

MMINLINE void MM_MarkingDelegate::workerCompleteGC(MM_EnvironmentBase *env)
{
		/* All threads flush buffers before this point, and complete any remaining language-specific marking tasks */
		if (env->_currentTask->synchronizeGCThreadsAndReleaseSingleThread(env, UNIQUE_ID)) {
			/* Perform single-threaded tasks here */
			env->_currentTask->releaseSynchronizedGCThreads(env);
		}
}

jason-hall replied:
Moving the root clearer code to markLiveObjectsRoots, as well, would change the original order of the code. I'm worried that will break it

amicic replied:
It does have it - not sure why you did not see it: https://github.com/eclipse/openj9/blob/master/runtime/gc_glue_java/MarkingDelegate.cpp#L158

amicic commented on runtime/gc_realtime/RealtimeMarkingScheme.cpp:
```diff
@@ -480,502 +83,89 @@ void
-	GC_SlotObject referentPtr(_javaVM->omrVM, &J9GC_J9VMJAVALANGREFERENCE_REFERENT(env, objectPtr));
+	MM_EnvironmentRealtime *realtimeEnv = MM_EnvironmentRealtime::getEnvironment(env);
+	/* Process reference objects and finalizable objects. */
+	_realtimeGC->getRealtimeDelegate()->markLiveObjectsScan(realtimeEnv);
```
This should be called earlier. Before the sync point. And the only thing it should call is flushReferenceObjects.

jason-hall replied:
Again, this would change the original order of the code and break it.

amicic replied:
it will not the break the order if you change what it does, it should be exactly at the spot what now flushReferenceObjects is called

amicic replied:
as I explained curently markLiveObjectsScan in delegate does the wrong thing. This is the main scan work not mean for any of root passes

jason-hall commented:
@amicic It seems your suggestions on how to break down the markLiveObjects() into phases means significantly rewriting the entire order of of the steps of the original code. I'm concerned that may break something? Also, the base class is only stub functions with no implementations. It doesn't really serve as an example.

amicic commented:
when I talk about base class I mean MarkingScheme in runtime/gc_glue_java. It's a parrent calss of RealtimeMarkingScheme. It's not a class with empty stub functions

amicic commented:
And similarly for the delegate you should look into MarkingDelegate in runtime/gc_glue/java, Don't look into example delegate in OMR - they don't have any language specific, that Realtime or standard Java Marking has.

amicic commented:
There is no significant rewrite -it's just an intorduction of 4 wrappers (well 8, for Scheme and Delegate) and moving the code from markLiveObjects to those wrappers . Order must not and will not be changed - it's not hard

jason-hall commented:
@amicic I didn't find it perfectly clear or maybe I misread something. I think I understand it correctly now with the additional details. Thank you.

I'm rewriting my phases to be more like the logic in the gc_glue_java/MarkingDelegate functions called by the stub parent base class MarkingScheme at omr/gc/MarkingScheme as suggested.

amicic commented:
looks about right after the last changes, but I'll have another fresh look

dmitripivkine commented on runtime/gc_realtime/RealtimeMarkingSchemeRootClearer.hpp:
```diff
@@ -0,0 +1,291 @@
+	/**
+	 * This should not be called
+	 */
+	MMINLINE virtual void
```
Declaration `MMINLINE` is ignored with `virtual` and can be removed... here and everywhere

dmitripivkine commented on runtime/gc_realtime/RealtimeMarkingSchemeRootClearer.hpp:
```diff
@@ -0,0 +1,291 @@
+		PORT_ACCESS_FROM_ENVIRONMENT(_env);
+		j9tty_printf(PORTLIB, "MM_RealtimeMarkingSchemeRootClearer::doSlot should not be called\n");
+		assert(false);
+	}
```
body of this function should be replaced to
```
{
    Assert_MM_unreachable();
}
```

jason-hall commented:
@dmitripivkine @amicic Is this done or did I miss anything?

dmitripivkine commented:
Jenkins compile win

dmitripivkine commented:
Jenkins test sanity xLinux jdk8

dmitripivkine commented:
Jenkins compile win jdk8

dmitripivkine commented:
Jenkins test sanity AIX jdk11

dmitripivkine commented:
Jenkins test sanity AIX jdk11
